src/Cardano/Shell/NodeIPC/Lib.hs

```diff
@@ -139,61 +161,112 @@ getIPCHandle = do
-    catches handleMsgIn [Handler handler, Handler handleMsgError]
+ipcListener :: forall m . (MonadIO m, MonadCatch m, MonadMask m) => ReadHandle -> WriteHandle -> Port -> m ()
+ipcListener readHandle@(ReadHandle rHndl) writeHandle@(WriteHandle wHndl) (Port port) =
+    finally
+        (do
+            checkHandles readHandle writeHandle
+            catches handleMsgIn [Handler handler, Handler handleMsgError]
+        )
+        shutdown
   where
     handleMsgIn :: m ()
     handleMsgIn = do
         liftIO $ hSetNewlineMode rHndl noNewlineTranslation
         send Started
-        msgIn <- readMessage readHndl
+        msgIn <- readMessage readHandle
         case msgIn of
-            QueryPort -> send (ReplyPort port) >> shutdown
-            Ping      -> do
-                send Pong
-                shutdown
+            QueryPort          -> send (ReplyPort port)
+            Ping               -> send Pong
             -- TODO:Handle them nicely
-            MessageInFailure _ -> shutdown
+            MessageInFailure _ -> return ()
 
     send :: MsgOut -> m ()
-    send = sendMessage writeHndl
+    send = sendMessage writeHandle
 
-    -- Huge catch all, fix this:
     handler :: IOError -> m ()
     handler err = do
-        liftIO $ do
-            when (isEOFError err) $ logError "its an eof"
-            hClose rHndl
-            hClose wHndl
-            hFlush stdout
+        liftIO $ when (isEOFError err) $ logError "its an eof"
         throwM IPCException
 
+    shutdown :: m ()
+    shutdown = liftIO $ do
+        hClose rHndl
+        hClose wHndl
+        hFlush stdout
+
     handleMsgError :: MessageException -> m ()
     handleMsgError err = do
         liftIO $ logError "Unexpected message"
         send $ MessageOutFailure $ ParseError $ show err
 
-    -- (TODO:) Exception handling on broken handles (e.g.handle is already closed etc.)
-    -- Implement here
-    shutdown :: m ()
-    shutdown = return ()
+    -- | Check if given two handles are usable (i.e. Handle is open, can be used to read/write)
+    checkHandles :: ReadHandle -> WriteHandle -> m ()
+    checkHandles (ReadHandle rHandle) (WriteHandle wHandle) = liftIO $ do
+        checkHandle rHandle hIsOpen (HandleClosed rHandle)
+        checkHandle wHandle hIsOpen (HandleClosed wHandle)
+        checkHandle rHandle hIsReadable (UnreadableHandle rHandle)
+        checkHandle wHandle hIsWritable (UnwritableHandle wHandle)
 
+    checkHandle :: Handle -> (Handle -> IO Bool) -> NodeIPCException -> IO ()
+    checkHandle handle pre exception = do
+        result <- pre handle
+        when (not result) $ throwM exception
 
 --------------------------------------------------------------------------------
--- placeholder
+-- Placeholder
 --------------------------------------------------------------------------------
 
+-- | Use this until we find suitable logging library
 logError :: Text -> IO ()
 logError _ = return ()
```

Comment:
What about simply outputting this message on `stdOut` or `stdErr` for now?

Reply:
At the end of the day, we will move these files into the executable anyway, so it's not really critical.